File: 503.next-greater-element-ii.go
func max(a int, b int) int {
	if a > b {
		return a
	}

	return b
}

func nextGreaterElements(nums []int) []int {
	var v []int
	if len(nums) == 0 {
		return v
	}

	maxn := nums[0]

	for i := 0; i < len(nums); i++ {
		maxn = max(maxn, nums[i])
	}

	for i := 0; i < len(nums); i++ {
		found := false

		if found == false {
			if maxn == nums[i] {
				v = append(v, -1)
				found = true
			}
		}

		if found == false {
			for j := i+1; j < len(nums); j++ {
				if nums[j] > nums[i] {
					v = append(v, nums[j])
					found = true
					break
				}
			}
		}

		if found == false {
			for j := 0; j < i; j++ {
				if nums[j] > nums[i] {
					v = append(v, nums[j])
					found = true
					break
				}
			}
		}

		if found == false {
			v = append(v, -1)
		}
	}

	return v
}

File: 503.next-greater-element-ii.cpp
// CreateTime: 2020-12-05 02:18:51
class Solution {
public:
    typedef pair<int, int> PII;

    vector<int> nextGreaterElements(vector<int>& nums) {
        int len = nums.size();

        deque<PII> que;
        vector<int> ans(len);
        for (int i = 2*len-1; i >= 0; i--) {
            auto x = nums[i%len];

            while (que.size() && (que.back().first-i+1 > len || x >= que.back().second)) {
                que.pop_back();
            }

            if (que.size() && i <= len-1) {
                ans[i] = que.back().second;
            } else if (i <= len-1) {
                ans[i] = -1;
            }

            que.push_back({i, x});
        }

        return ans;
    }
};
